### ISSUE TITLE
Netty with accessEvents with logback-access

### ISSUE THREAD

**Issue by bhochhi** [reactor/reactor-netty#781]
Why netty has no accessEvents but just accessLog implementation? Is there any plan supporting that?

**Comment by violetagg** [comment]
Such integration should be possible via https://projectreactor.io/docs/netty/release/api/reactor/netty/http/server/HttpServer.html#accessLog-boolean-reactor.netty.http.server.logging.AccessLogFactory-

Marking this as `help wanted` if somebody wants to work on this.

**Comment by childewuque** [comment]
Thanks
It would be better if we could configure instead of coding

**Comment by JooHyukKim** [comment]
has anyone worked on this yet?

**Comment by violetagg** [comment]
@JooHyukKim Currently nobody is working on this one.

**Comment by JooHyukKim** [comment]
@violetagg I see. Would you allow me to work on it? seems quite interesting

**Comment by JooHyukKim** [comment]
@violetagg @bhochhi I'd like to hear more about your thoughts on what AccessEvents are and how they should behave.

**Comment by violetagg** [comment]
> @violetagg I see. Would you allow me to work on it? seems quite interesting

@JooHyukKim Yes sure

**Comment by violetagg** [comment]
> @violetagg @bhochhi I'd like to hear more about your thoughts on what AccessEvents are and how they should behave.

I believe that @bhochhi meant integration with `logback-access` https://logback.qos.ch/access.html

**Comment by JooHyukKim** [comment]
trueeee. @bhochhi is that what you are suggesting? integration with `logback-access` like how Tomcat and Jetty are integrated with `logback-access` described in [https://logback.qos.ch/access.html](https://logback.qos.ch/access.html) ?

**Comment by JooHyukKim** [comment]
`logback-access` is officailly described to [integrate with Serverlet Containers such as Jetty or Tomcat](https://logback.qos.ch/access.html#:~:text=integrates%20with%20Servlet%20containers%20such%20as%20Jetty%20or%20Tomcat). 
As far as I know, we also need **netty-specific** `Adaptor` and `ValveBase` implementation in `Logback-access`.
looking at [codebase of Tomcat and Jetty integration with logback-access](https://github.com/qos-ch/logback/tree/master/logback-access/src/main/java/ch/qos/logback/access).

Lemme know how you think :)
cheers

**Comment by bhochhi** [comment]
yes, I wanted Netty to support AccessEvent so that logback-access can be supported. However, my original question is Why Netty didn't have support? Is that by design or other limitations due to different paradigms?

**Comment by violetagg** [comment]
> yes, I wanted Netty to support AccessEvent so that logback-access can be supported. However, my original question is Why Netty didn't have support? Is that by design or other limitations due to different paradigms?

I don't understand this.

In Tomcat you have Valve concept, here in Reactor Netty you have a mechanism for a custom implementation for the access logging https://projectreactor.io/docs/netty/release/api/reactor/netty/http/server/HttpServer.html#accessLog-boolean-reactor.netty.http.server.logging.AccessLogFactory-

**Comment by JooHyukKim** [comment]
I think this issue needs more clarification on exactly what the desired enhancement is 👍

**Comment by childewuque** [comment]
@JooHyukKim 
https://github.com/reactor/reactor-netty/issues/1752

For example, add remote host in logback.xml or in other file,like nginx format
Thanks

**Comment by JooHyukKim** [comment]
> Such integration should be possible via https://projectreactor.io/docs/netty/release/api/reactor/netty/http/server/HttpServer.html#accessLog-boolean-reactor.netty.http.server.logging.AccessLogFactory-
> 
> Marking this as `help wanted` if somebody wants to work on this.

@childewuque
kindly suggest specific purpose and way to add the functionality? thanks.

**Comment by dmitrysulman** [comment]
@violetagg I encountered this issue while investigating ways to configure Reactor Netty access logs more flexibly.

From my POV, the fundamental limitation preventing integration with `logback-access` or other tools is that [AccessLog](https://github.com/reactor/reactor-netty/blob/ddb27af3c034a0086ddd13a4aa27068c995120e1/reactor-netty-http/src/main/java/reactor/netty/http/server/logging/AccessLog.java) is a `final` class.

Currently, the only way to customize access logs is by modifying the `reactor.netty.http.server.AccessLog` logger. However, this approach doesn’t allow integration with `logback-access`, since `logback-access` uses a different type of logging event than standard logs.

If `AccessLog` were an interface, it would be possible to provide a custom [AccessLogFactory](https://github.com/reactor/reactor-netty/blob/ddb27af3c034a0086ddd13a4aa27068c995120e1/reactor-netty-http/src/main/java/reactor/netty/http/server/logging/AccessLogFactory.java) that creates an `AccessLog` implementation with a custom `log()` method enabling seamless integration. For example:

```java
public class MyAccessLog implements AccessLog {

	private final SomeExternalAccessLogHandler handler;
	private final AccessLogArgProvider argProvider;

	public MyAccessLog(SomeExternalAccessLogHandler handler, AccessLogArgProvider argProvider) {
		this.handler = handler;
		this.argProvider = argProvider;
	}

	@Override
	public void log() {
		handler.handle(argProvider);
	}

}
```

I've created a small POC branch introducing this change. If this approach looks good to you, I'd be happy to open a PR:

https://github.com/reactor/reactor-netty/compare/main...dmitrysulman:reactor-netty:gh-781

**Comment by violetagg** [comment]
@dmitrysulman We cannot make backwards incompatible changes. However I'll try to find time to look at this feature in more details.

**Comment by dmitrysulman** [comment]
@violetagg thank you!

Maybe I'm missing something but I don't see how this change can break backward compatibility. The only `public` method in the `AccessLog` class is `public static AccessLog create(String logFormat, Object... args)` and that remains unchanged. The constructor is `private`, so the class can't be instantiated directly.

The only case I can think of that might break is if someone uses reflection to call the private constructor. But as far as I know, that's not considered part of the public API. Please correct me if I'm wrong.

The simpler alternative would be to just remove the `final` modifier from the class and make the `log()` method `public`.

**Comment by dmitrysulman** [comment]
> @dmitrysulman Would you like to propose this as PR?

@violetagg sure. Just to clarify, you're talking about a PR that removes the `final` modifier, right? And which branch should it target, `main` or `1.2.x`?

**Comment by violetagg** [comment]
@dmitrysulman yes, PR that removes the `final` modifier. You can target the branch where you need it either `1.2.x` or `main` (`main` is the current development and will be released in November)

**Comment by dmitrysulman** [comment]
@violetagg here is the PR: #3726

**Comment by dmitrysulman** [comment]
@violetagg thank you for merging the PR.

I encountered another small issue that prevents the integration with external tools such as `logback-access`. The problem is that [AccessLogArgProvider](https://github.com/reactor/reactor-netty/blob/d8c9d11d6c08d5e8615e35a6b83ccd2d67b6c884/reactor-netty-http/src/main/java/reactor/netty/http/server/logging/AccessLogArgProvider.java) doesn't provide methods for retrieving the full lists of request and response headers. While it has methods like `requestHeader(CharSequence name)` and `responseHeader(CharSequence name)`, this isn't sufficient because the `logback-access` [IAccessEvent](https://github.com/qos-ch/logback-access/blob/70b7e37d0e7b995ee142cc154109fed8378d0164/logback-access-common/src/main/java/ch/qos/logback/access/common/spi/IAccessEvent.java) interface exposes `getRequestHeaderMap()` and `getResponseHeaderMap()` methods and some logic requires iterating over the complete list of headers.

For example, in [FullRequestConverter#convert](https://github.com/qos-ch/logback-access/blob/70b7e37d0e7b995ee142cc154109fed8378d0164/logback-access-common/src/main/java/ch/qos/logback/access/common/pattern/FullRequestConverter.java#L35-L51):
```java
    public String convert(IAccessEvent ae) {
        StringBuilder buf = new StringBuilder();
        buf.append(ae.getRequestURL());
        buf.append(CoreConstants.LINE_SEPARATOR);


        Enumeration<String> headerNames = ae.getRequestHeaderNames();
        while (headerNames.hasMoreElements()) {
            String name = headerNames.nextElement();
            buf.append(name);
            buf.append(": ");
            buf.append(ae.getRequestHeader(name));
            buf.append(CoreConstants.LINE_SEPARATOR);
        }
        buf.append(CoreConstants.LINE_SEPARATOR);
        buf.append(ae.getRequestContent());
        return buf.toString();
    }
```

Fortunately, all Netty HTTP headers interfaces (`HttpHeaders`, `Http2Headers`, `Http3Headers`) expose appropriate request/response header `Iterator` methods that could be easily integrated into all `AccessLogArgProvider` implementations to support this need.

I've prepared a PR that addresses this enhancement. Please let me know if it needs any improvements.

**Comment by oemergenc** [comment]
@dmitrysulman i would be really interested to see your full solution. Any changes you publish a small demo project?

**Comment by dmitrysulman** [comment]
@oemergenc Sure, I'm in the process of developing a library that will enable first-class integration between the Reactor Netty HTTP Server and Logback Access. It is almost ready, here is the repository: https://github.com/dmitrysulman/logback-access-reactor-netty.

Currently, it works with Reactor Netty `1.2.6-SNAPSHOT`. I'm planning to release it and publish it to Maven Central after the Reactor Netty `1.2.6` release which will include all the necessary enhancement discussed in this task.

The future plan is to develop Spring Boot starter with auto-configuration.

**Comment by oemergenc** [comment]
Wow, this is great. Looking forward for the first release.

**Comment by dmitrysulman** [comment]
@oemergenc 
The library is released, you can give it a try: https://github.com/dmitrysulman/logback-access-reactor-netty.

As I already mentioned this library introduces first-class integration between **Reactor Netty HTTP Server** and **Logback Access** for access logging. Configuration is very simple, you just need to provide standard Logback Access `.xml` configuration file (like `logback-access.xml`) and that's it.

Spring Boot starter with auto-configuration is coming soon.

Any feedback is more than welcome!

**Comment by oemergenc** [comment]
@dmitrysulman the library looks nice. It would be awesome if this could be integrated into the corresponding spring project. I think this is missing for a very long time.

**Comment by dmitrysulman** [comment]
@oemergenc Spring Boot starter [is also released](https://github.com/dmitrysulman/logback-access-reactor-netty?tab=readme-ov-file#using-as-a-spring-boot-starter).

Feedback is welcome!
